### ISSUE TITLE
Flatpak build

### ISSUE THREAD

**Issue by LeandroStanger** [BleuBleu/FamiStudio#41]
Please, consider supporting [Flatpak](http://docs.flatpak.org/en/latest/introduction.html#reasons-to-use-flatpak) and publishing FamiStudio on [Flathub](https://github.com/flathub/flathub/wiki/App-Submission).

**Comment by BleuBleu** [comment]
Hi. 

Some other people have asked for FamiStudio to come in the their favorite package format (Snap/AUR, FlatPak, etc.).

The Linux community is very small at the moment (the download numbers are **extremely** low), so id rather invest my energy elsewhere at the moment until the community grows up a bit. Example in hand, I literally am not aware of a single song made with the Linux version at the moment! 😂 My impression is that Linux users only seem to download the app, launch it, and close it.

I need to pick my battles, I hope you understand.

-Mat

**Comment by ernstp** [comment]
I'm trying to put something together... :-)

**Comment by ernstp** [comment]
Ok I put something together.. @BleuBleu do you care to take a look? It says BleuBleu everywhere so I won't do anything without your approval:

https://github.com/ernstp/flathub/tree/new-pr

Here's the process: https://github.com/flathub/flathub/wiki/App-Submission

If you want to own the submission feel free to copy it. I'm thinking about the "Someone else has put my app on Flathub—what do I do?" section.

**Comment by ernstp** [comment]
Here's a test build:
flatpak install --user https://dl.flathub.org/build-repo/26142/org.BleuBleu.FamiStudio.flatpakref

**Comment by BleuBleu** [comment]
Hi! 

Sorry for the delay, these releases take a lot out of me so I tend to take a short GitHub break after each one. I just tested the test build and it worked spectacularety well on the VM i setup!

Now as you might have guessed, I am a Linux-moron, so bear with me while I ask you the **dumbest** questions:

1. I noticed it installed a bunch of dependencies, Mono was not one of them. Was it because I already had it? I do see it in the YML file mentionned a bunch.
2. If I integrate this, what do I need to do at each release to publish the build on flathub? What is the exact series of action i need to perform.
3. I see there is a SVG icon in there, its super small. In the FamiStudio.svg file in the Resource folder, there are actually higher-res versions of the logo outside of the viewport (see screenshot i attached). Could we switch to these instead?
4. If users have issues with the package, would you be willing to minimally support them at the start? As I said, the time i have to devote to Linux is very limited at the moment. 

Thanks again for the contribution, and I hope we can get the conversation going and get that released!

![Logos](https://user-images.githubusercontent.com/6654334/93017483-b4f87680-f5fb-11ea-892e-d261f04530db.JPG)

-Mat

**Comment by ernstp** [comment]
Hi! Nice that you liked it! It gives quite good visibility and convenience for an app.

Here's the pull request for FlatHub: https://github.com/flathub/flathub/pull/1817

1) There's no Mono "runtime" so it's just bundled inside the app package itself. But there is an "sdk" which is official which is used to add Mono support to the app.
It's the combination of
```
sdk-extensions:
  - org.freedesktop.Sdk.Extension.mono6
```
and:
```
  - name: mono6
    buildsystem: simple
    build-commands:
      - /usr/lib/sdk/mono6/install.sh
```

2) So if you just integrate it here you can build a Flatpak yourself etc, but I think if we want it published on FlatHub specifically they want a special FlatHub app repo there. I get that impression from their App Submission guidelines. I can ask for more details.
They have one repo for every app like this: https://github.com/flathub/com.microsoft.Teams
I can make sure you're admin of that repo also (I assume!).

3) Right, I picked the SVG icon since it's scaleble and works at any size. However I see that it has less detail than the others now, I'll switch.

4) Sure!

**Comment by bilelmoussaoui** [comment]
Hey, I am a Flathub reviewer. Flathub uses a github repository under https://github.com/flathub per app identifier. The default branch of the repository is used to build the stable release, applications developers can use the beta branch as well if they want.

Whenever a PR is created against the repo, the buildbot will build the manifest and provide with you a test build you can easily install, run & test. Once the changes looks good, you can merge it, a new build type will take place that will publish the update to users few hours after that (it's set to 3 hours by default which you can change).

**Comment by ernstp** [comment]
Oh I didn't see your edit, I will update to the new SVG!

**Comment by ernstp** [comment]
Looks pretty neat, eh? https://flathub.org/apps/details/org.famistudio.FamiStudio

**Comment by BleuBleu** [comment]
Hello hello! Again sorry for the delay, catching up on stuff now!

Wow, it looks all legit now! :) Amazing work!!! You just made a lot of people super happy!

So early next week I will likely release 2.2.1 to fix a small issue.

Just to confirm that I understood correctly :

1) I simply package everything like I did before. 
2) I update the package at https://github.com/flathub/org.famistudio.FamiStudio with new version, links, release notes, etc.
3) Create a PR and it should get picked up auto-magically.

That's it?

-Mat

**Comment by ernstp** [comment]
1) Yup.
2) Yup.
3) Yeah, you also got push access to it right? So we can just merge stuff on our own. And yes, it should be built automatically...

**Comment by ernstp** [comment]
You could add a longer summary to the metainfo.xml file.

**Comment by BleuBleu** [comment]
Excellent, yes! I got an invite to push directly in the repo I think. I'll keep the issue open until I release 2.2.1 in case I have other dumb questions! :)

Once i do it once I should feel a bit more confident.

-Mat

**Comment by BleuBleu** [comment]
Hi! I was wondering if I could ask you another general Linux question. 

One of the issues a lot of people have with the way I distribute the app on Linux, is that I distribute binaries of stuff that I compiled locally. The 2 libs : OpenAL & rtMIDI, as well as 2 C++ DLLs : NesSndEmu and NotSoFatso. 

These assumes people have the required dependencies installed. To minimize this kind of issue, I link with static CRT (which is why they are so large) and compile everything with a ridiculously old version of gcc. Still, every now and then some people running old/exotic distros will be missing some required dependencies.

It its current form, is the Flathub build less likely to have this kind of issue? It is still relying on my locally compiled files afterall.

-Mat

**Comment by ernstp** [comment]
Right I mean this is the kinds of problems flatpak etc are meant to solve.
The app runs in an environment controlled by the flatpak, so it is controlled by the publisher and should work the same for everyone.

All the dependencies should be bundled with the app or come from the selected runtime. So you can see we use the Gnome 3.36 runtime here: https://github.com/flathub/org.famistudio.FamiStudio/blob/07d95aab65a35a020c1086dc670db08b2505467c/org.famistudio.FamiStudio.yml#L3

You can see here that I threw out your provided libopenal and replaced it with a symlink from the runtime:
https://github.com/flathub/org.famistudio.FamiStudio/blob/07d95aab65a35a020c1086dc670db08b2505467c/org.famistudio.FamiStudio.yml#L48

Stuff that's not in a runtime you have to copy into the app itself.

**Comment by mcgrew** [comment]
First, let me say that I love this app. I just downloaded it yesterday, but I was able to export one of the demos and compile it into an NES game that I'm working on. Already much more useful than any other solution I've found. The only problem is that I'm terrible at music composition but I'm going to try to learn.

On the topic of Linux packaging, I might be willing to maintain a package in the Archlinux User Repository (AUR). I have a few applications of my own that I keep updated there, so I know how it works. Flatpak is fine and should be supported, but some people (like myself) prefer not to use flatpak, so an AUR package would provide a way to manage the installation with the system's package manager on Arch (and other arch-based distributions like Manjaro).

If you're alright with that I will look into proper packaging for Arch that will handle any dependencies the application should need.

I may also have some suggestions/PRs for optimizing the sound engine in the near future since I've been messing with it some.
